src/Http/Routing/src/EndpointMiddleware.cs

```diff
@@ -52,6 +60,34 @@ public async Task Invoke(HttpContext httpContext)
             await _next(httpContext);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void EnsureRequisiteMiddlewares(HttpContext httpContext, Endpoint endpoint)
+        {
+            if (endpoint.Metadata.GetMetadata<IAuthorizeData>() != null &&
+                !httpContext.Items.ContainsKey(AuthorizationMiddlewareInvokedKey))
+            {
+                ThrowMissingAuthMiddlewareException(endpoint);
+            }
+
+            if (endpoint.Metadata.GetMetadata<ICorsMetadata>() != null &&
+                !httpContext.Items.ContainsKey(CorsMiddlewareInvokedKey))
+            {
+                ThrowMissingCorsMiddlewareException(endpoint);
+            }
+        }
+
+        private static void ThrowMissingAuthMiddlewareException(Endpoint endpoint)
+        {
+            throw new InvalidOperationException($"Endpoint {endpoint.DisplayName} contains authorization metadata, " +
+                "but a middleware was not found that supports authorization.");
```

Comment:
We say "a middleware", but we really require middleware that sets `AuthorizationMiddlewareInvokedKey`.

Is this a problem for people who think that they can implement their own middleware to fix this problem? I think this exception could explicitly say `.UseAuthorization()` is required. It is the only middleware that knows to set `AuthorizationMiddlewareInvokedKey`

The alternative is providing a more first class way of saying that auth has run for the request.

Reply:
The error message is a bit more explicit about what to do resolve it and includes a line saying you could add auth. 

> The alternative is providing a more first class way of saying that auth has run for the request.

We could add a gesture for this or turn it in to a feature with settable properties. Either of these seem viable to me